File: csharp-playwright-nunit-reqnroll/SauceDemo.UiTests/Support/AllureAttachments.cs
using System.Text;
using Allure.Net.Commons;
using NUnit.Framework;
using SauceDemo.UiTests.Drivers;

namespace SauceDemo.UiTests.Support;

/// <summary>Attaches failure evidence to the current Allure test result. Allure.Net.Commons uses AsyncLocal for its lifecycle context, so this is safe to call concurrently from parallel scenarios.</summary>
public static class AllureAttachments
{
    /// <summary>
    /// Collects whatever failure evidence the browser can still give us. Each attachment is captured
    /// independently and best-effort, in ascending order of how much of a live browser it needs,
    /// because the scenarios that most need evidence include the ones that failed *because* the page
    /// or browser died - and there a screenshot throws while <see cref="Microsoft.Playwright.IPage.Url"/>
    /// (a locally cached value, no IPC) still answers.
    ///
    /// Nothing here is allowed to throw. This runs from an [AfterScenario] hook, and Reqnroll stops
    /// firing the remaining [AfterScenario] hooks as soon as one throws - so an exception escaping here
    /// would skip the hooks that close the browser and return this scenario's pooled login user,
    /// exhausting the user pool and making every later scenario block until UserAcquireTimeoutSeconds.
    /// Losing a screenshot is a far better outcome than losing the rest of the run.
    /// </summary>
    public static async Task AttachFailureEvidenceAsync(BrowserSession session)
    {
        await TryAttachAsync("final-url", "text/plain", ".txt",
            () => Task.FromResult(Encoding.UTF8.GetBytes(session.Page.Url)));

        await TryAttachAsync("failure-screenshot", "image/png", ".png",
            () => ScreenshotHelper.CaptureAsync(session.Page));

        await TryAttachAsync("page-source", "text/html", ".html",
            async () => Encoding.UTF8.GetBytes(await session.Page.ContentAsync()));

        // Best-effort only, like the Selenium sibling's console-log attachment: unlike Selenium's
        // pull-based Manage().Logs API, Playwright's console log is push-based (BrowserSession
        // subscribes to IPage.Console as soon as the page is created), so there's nothing to catch here
        // beyond "were there any messages at all".
        if (!session.ConsoleLog.IsEmpty)
        {
            await TryAttachAsync("browser-console-log", "text/plain", ".txt",
                () => Task.FromResult(Encoding.UTF8.GetBytes(string.Join(Environment.NewLine, session.ConsoleLog))));
        }
    }

    private static async Task TryAttachAsync(string name, string mimeType, string extension, Func<Task<byte[]>> capture)
    {
        try
        {
            AllureApi.AddAttachment(name, mimeType, await capture(), extension);
        }
        catch (Exception ex)
        {
            // Deliberately catches everything - see the contract on AttachFailureEvidenceAsync. The
            // note goes to the test's own captured output so a missing attachment is explainable
            // rather than mysterious.
            TestContext.Out.WriteLine($"Could not attach '{name}' failure evidence: {ex.GetType().Name}: {ex.Message}");
        }
    }
}

File: ts-playwright-cucumber/src/support/allureAttachments.ts
import * as allure from 'allure-js-commons';
import type { ScenarioSession } from '../browser/browserManager.ts';
import { captureScreenshot } from './screenshotHelper.ts';

/**
 * Collects whatever failure evidence the browser can still give us. Each attachment is captured
 * independently and best-effort, in ascending order of how much of a live page/browser it needs,
 * because the scenarios that most need evidence include the ones that failed *because* the page or
 * browser died - and there a screenshot/content read throws while `page.url()` (a plain, synchronous,
 * locally-cached read - not even async, unlike every other method here that touches the live page)
 * still answers.
 *
 * Nothing here is allowed to throw: this runs from an `After` hook, and while Cucumber.js (unlike
 * Reqnroll) does not abort later same-type hooks when one throws, a failure here still shouldn't cost
 * the run its actual evidence-gathering attempts for the OTHER attachments below it.
 */
export async function attachFailureEvidence(session: ScenarioSession): Promise<void> {
  await tryAttach('final-url', 'text/plain', () => Promise.resolve(session.page.url()));
  await tryAttach('failure-screenshot', 'image/png', () => captureScreenshot(session.page));
  await tryAttach('page-source', 'text/html', () => session.page.content());

  // Best-effort only, like the other siblings' console-log attachment: unlike Selenium's pull-based
  // Manage().Logs API, the console log here is push-based (browserManager.ts subscribes to the page's
  // 'console' event as soon as the page is created), so there's nothing to catch here beyond "were there
  // any messages at all".
  if (session.consoleLog.length > 0) {
    await tryAttach('browser-console-log', 'text/plain', () =>
      Promise.resolve(session.consoleLog.join('\n'))
    );
  }
}

async function tryAttach(
  name: string,
  mimeType: string,
  capture: () => Promise<string | Buffer>
): Promise<void> {
  try {
    await allure.attachment(name, await capture(), mimeType);
  } catch (error) {
    const reason = error instanceof Error ? `${error.name}: ${error.message}` : String(error);
    console.error(`Could not attach '${name}' failure evidence: ${reason}`);
  }
}
